### PR TITLE
[StorageQuota] Adding EnforcementPolicy to enforce storage quota based on clustermap

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
1. Add a new interface StorageQuotaEnforcementPolicy
2. Add two implementations of such interface
    2.1 A simple enforcement policy, once the current usage exceeds storage quota, throttle the request
    2.2 An adaptive enforcement policy, who also takes cluster capacity into consideration. If the cluster has more than 25% of partitions are writable, don't throttle.

### CODE REVIEW

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/AdaptiveStorageQuotaEnforcementPolicyFactory.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  public AdaptiveStorageQuotaEnforcementPolicyFactory(ClusterMap clusterMap, StorageQuotaConfig config) {
+    policy = new AdaptiveStorageQuotaEnforcementPolicy(clusterMap, SystemTime.getInstance(), config);
+  }
+
+  @Override
+  public StorageQuotaEnforcementPolicy getPolicy() {
+    return policy;
+  }
+}
+
```
Comment (tbingol) [inline ambry-quota/src/main/java/com/github/ambry/quota/storage/AdaptiveStorageQuotaEnforcementPolicyFactory.java:42]: Can you add some comments about this implementation?

File: ambry-quota/src/main/java/com/github/ambry/quota/storage/AdaptiveStorageQuotaEnforcementPolicyFactory.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        updateClusterMapWritablePartitionPercentage();
+      }
+    }
+    return currentUsage >= storageQuota
+        && currentClusterMapWritablePartitionPercentage <= config.adaptiveEnforcementWritablePartitionThreshold;
+  }
+
+  private void updateClusterMapWritablePartitionPercentage() {
+    this.currentClusterMapWritablePartitionPercentage =
+        100 * clusterMap.getWritablePartitionIds(null).size() / clusterMap.getAllPartitionIds(null).size();
```
Comment (tbingol) [inline ambry-quota/src/main/java/com/github/ambry/quota/storage/AdaptiveStorageQuotaEnforcementPolicyFactory.java:72]: Though it should not happen, we better make sure to handle divide by zero case

File: ambry-api/src/main/java/com/github/ambry/config/StorageQuotaConfig.java [hunk 3]
```diff
@@ -102,6 +109,28 @@
...
+
+  /**
+   * Interval to update clustermap data in adaptive enforcement policy.
+   */
+  @Config(ADAPTIVE_ENFORCEMENT_CLUSTERMAP_UPDATE_INTERVAL_MS)
+  @Default("10 * 60 * 1000")
+  public final long adaptiveEnforcementClusterMapUpdateIntervalMs;
+
+  /**
+   * Threshold in percentage when percentage of writable partitions drops below this threshold, we adaptive enforcement
```
Comment (tbingol) [inline ambry-api/src/main/java/com/github/ambry/config/StorageQuotaConfig.java:127]: Nit: s/we adaptive/the adaptive

**Review by tbingol — APPROVED** [review]

**Review by ankagrawal — COMMENTED** [review]
I don't think its a good idea to allow storage above quota. Storage once used is not a resource we can get back later. If, for example, due to a bug, a customer writes a lot of data to an account, and we allow the write to then continue on free space, then it will be a lot of operational effort for us to get that storage back. Imagine a case where writes due to bugs are also intermingled with legitimate writes. It will be operationally very time taking to figure out all the blobs that were erroneously written and then delete them , and then wait for compaction to catchup.

Even if we do decide to do this, we should do this only on a fraction "unpromised storage capacity" available. This will be the storage capacity of a cluster that is extra and not part of any customer's quota. Note that even with this, since our storage quota is per month, it will be hard to come up with an estimate of how much we can give out above quota.

So for storage quota, I think the cleanest approach will be to allocate storage quota generously, give out warnings and alerts proactively, but not allow customers to exceed quota.

**Pull request closed without merging**